lib/jsonapi/query_builder/mixins/sort.rb

```diff
@@ -28,45 +28,95 @@ def default_sort(options)
             @_default_sort = options
           end
 
-          def sorts_by(*attributes)
-            @_sort_attributes = _sort_attributes + attributes
+          def sorts_by(*attributes, **nested_attributes)
+            nested_sort_attributes = _sort_attributes.extract_options!
+            @_sort_attributes = _sort_attributes + attributes + [nested_sort_attributes.deep_merge(nested_attributes)]
           end
         end
 
         def sort(collection, sort_params = send(:sort_params))
+          ensure_permitted_sort_params!(sort_params) if sort_params
+
           collection
-            .reorder(sort_params.nil? ? self.class._default_sort : formatted_sort_params(sort_params))
+            .tap { |c| add_order_attributes(c, sort_params) }
+            .tap { |c| add_joins_for_nested_attributes(c, sort_params) if sort_params }
             .tap(&method(:add_unique_order_attributes))
         end
 
         private
 
+        attr_reader :collection
+
         def sort_params
           params[:sort]
         end
 
+        def add_order_attributes(collection, sort_params)
+          collection.reorder(sort_params.nil? ? self.class._default_sort : formatted_sort_params(sort_params))
+        end
+
+        def add_joins_for_nested_attributes(collection, sort_params)
```

Comment:
I'm not sure if we should be doing "implicit" joins in the bowels of the gem, especially left ones. This seems like it could affect the result set (e.g. return duplicates). We have some examples on Underline where we have explicit joins or includes:

```ruby
def scope
  ResourceActiveUser.includes(:user)
end

def default_sort
  'users.first_name asc'
end
```

What would a sensible default be and how would we allow the users to change this joining logic?

Reply:
You're correct, this would work only for `belongs_to`/`has_one` relationships. But would sorting by a `has_many` relationship's attribute even make sense? Probably not.

Btw, look at the implicit joins AR makes with using includes and order:
<img width="627" alt="image" src="https://user-images.githubusercontent.com/16187935/93991131-3aaebb80-fd8c-11ea-8f33-918047be90c9.png">

There was a constant question in my head, is this even worth doing? Maybe we should opt for some kind of sort object, sorta (pun teheee 😄 ) like filter objects, so you could even use it for computed attributes.

Reply:
I agree, sorting by a `has_many` probably doesn't make sense. Sorting by a `belongs_to` shouldn probably be an exceptional case. We should probably push users of this gem into sorting only on the endpoint resource, and not on another resource connected to it. We probably need this for feature parity, so I think its fine if we go with it as an "experimental" feature until we get a better sense of what to do 👍

Reply:
Is "nested" the right word here, or could we change it to relationship? Nested reminds me of the whole nested_attributes thing.

I like the idea of a sort object, something similar to `filter` lambdas and objects.
Sorting by a relationship's attribute is rare, maybe it would be best to cover that via some type of sort objects or lambdas, to keep the gem simpler and the usage more explicit (without hidden joins).